public/app/plugins/datasource/graphite/add_graphite_func.ts

```diff
@@ -0,0 +1,159 @@
+import angular from 'angular';
+import _ from 'lodash';
+import $ from 'jquery';
+import rst2html from 'rst2html';
+import Drop from 'tether-drop';
+
+export function graphiteAddFunc($compile) {
+  var inputTemplate =
+    '<input type="text"' + ' class="gf-form-input"' + ' spellcheck="false" style="display:none"></input>';
+
+  var buttonTemplate =
+    '<a class="gf-form-label query-part dropdown-toggle"' +
+    ' tabindex="1" gf-dropdown="functionMenu" data-toggle="dropdown">' +
+    '<i class="fa fa-plus"></i></a>';
+
+  return {
+    link: function($scope, elem) {
+      var ctrl = $scope.ctrl;
+
+      var $input = $(inputTemplate);
+      var $button = $(buttonTemplate);
+
+      $input.appendTo(elem);
+      $button.appendTo(elem);
+
+      ctrl.datasource.getFuncDefs().then(function(funcDefs) {
+        var allFunctions = _.map(funcDefs, 'name').sort();
+
+        $scope.functionMenu = createFunctionDropDownMenu(funcDefs);
+
+        $input.attr('data-provide', 'typeahead');
+        $input.typeahead({
+          source: allFunctions,
+          minLength: 1,
+          items: 10,
+          updater: function(value) {
+            var funcDef = ctrl.datasource.getFuncDef(value);
+            if (!funcDef) {
+              // try find close match
+              value = value.toLowerCase();
+              funcDef = _.find(allFunctions, function(funcName) {
+                return funcName.toLowerCase().indexOf(value) === 0;
+              });
+
+              if (!funcDef) {
+                return '';
```

Comment:
Old one returned undefined here, i.e. `return;`. However it returned empty string at the end of update function :smile:  Can't see any reason why undefined or empty string should be returned, just want to make sure that your change from undefined to empty string was intentional.

Reply:
It was intentional. Typescript didn't like that it didn't return a value.